akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2Demux.scala

```diff
@@ -357,21 +360,39 @@ private[http2] abstract class Http2Demux(http2Settings: Http2CommonSettings, ini
         }
       })
 
+      // -----------------------------------------------------------------
       // FIXME: What if user handler doesn't pull in new substreams? Should we reject them
-      //        after a while or buffer only a limited amount? We should also be able to
-      //        keep the buffer limited to the number of concurrent streams as negotiated
-      //        with the other side.
+      //        after a while or buffer only a limited amount?
       val bufferedSubStreamOutput = new BufferedOutlet[Http2SubStream](substreamOut)
       override def dispatchSubstream(initialHeaders: ParsedHeadersFrame, data: Source[Any, Any], correlationAttributes: Map[AttributeKey[_], _]): Unit =
         bufferedSubStreamOutput.push(Http2SubStream(initialHeaders, data, correlationAttributes))
 
+      // -----------------------------------------------------------------
+      // This is a manually crafted sequence of shutdowns replacing the one on `completeStage`
+      // so we use the custom bufferedSubStreamOutput.complete() to ensure the buffer is empty
+      // before completing the outlet
+      override def onComplete(): Unit = {
+        cancel(substreamIn)
+        cancel(frameIn)
+        complete(frameOut)
+        bufferedSubStreamOutput.complete()
```

Comment:
`completeStage` would invoke:

```
        cancel(substreamIn)
        cancel(frameIn)
        complete(frameOut)
        complete(substreamOut)
```
which is wrong because it doesn't flush the buffer in `bufferedSubStreamOutput` in from of `substreamOut`.

Reply:
A matter of taste but I would have skipped the three liner comment since some of it is implicit by overriding and maybe rather added one above the `bufferedSubStreamOutput` saying something like `// important to delegate to bufferedSubStreamOutput complete(substreamOut)` since that is the only thing that is "manually crafted"